File: core-extra/arazzo-parser/src/main/java/org/evomaster/arazzo/deserializer/SuccessReusableDeserializer.java
package org.evomaster.arazzo.deserializer;

import com.fasterxml.jackson.core.JacksonException;
import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import org.evomaster.arazzo.models.domain.Reusable;
import org.evomaster.arazzo.models.domain.SuccessAction;
import org.evomaster.arazzo.models.domain.SuccessReusable;

import java.io.IOException;

public class SuccessReusableDeserializer extends JsonDeserializer<SuccessReusable> {

    @Override
    public SuccessReusable deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JacksonException {
        JsonNode node = jsonParser.getCodec().readTree(jsonParser);

        if (node.has("reference")) {
            Reusable reusable = jsonParser.getCodec().treeToValue(node, Reusable.class);
            return new SuccessReusable.ReusableObj(reusable);
        }

        SuccessAction action = jsonParser.getCodec().treeToValue(node, SuccessAction.class);
        return new SuccessReusable.Success(action);
    }

}

File: core/src/main/kotlin/org/evomaster/core/problem/rest/arazzo/deserializer/SuccessReusableDeserializer.kt
package org.evomaster.core.problem.rest.arazzo.deserializer

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind.DeserializationContext
import com.fasterxml.jackson.databind.JsonDeserializer
import com.fasterxml.jackson.databind.JsonNode
import org.evomaster.core.problem.rest.arazzo.models.Reusable
import org.evomaster.core.problem.rest.arazzo.models.SuccessAction
import org.evomaster.core.problem.rest.arazzo.models.SuccessReusable

class SuccessReusableDeserializer : JsonDeserializer<SuccessReusable>() {
    override fun deserialize(
        p0: JsonParser,
        p1: DeserializationContext
    ): SuccessReusable? {
        val node: JsonNode = p0.codec.readTree(p0)

        return if (node.has("reference")) {
            val reusable = p0.codec.treeToValue(node, Reusable::class.java)
            SuccessReusable.ReusableObj(reusable)
        } else {
            val action = p0.codec.treeToValue(node, SuccessAction::class.java)
            SuccessReusable.Inline(action)
        }
    }
}